tt_metal/hw/inc/debug/ring_buffer.h

```diff
@@ -0,0 +1,35 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#if defined(WATCHER_ENABLED)
+
+#include "hostdevcommon/debug_ring_buffer_common.h"
+
+// Returns the buffer address for current thread+core. Differs for NC/BR/ER/TR0-2.
+inline uint32_t* get_debug_ring_buffer() {
+#if defined(COMPILE_FOR_ERISC)
+    return reinterpret_cast<uint32_t*>(eth_l1_mem::address_map::ERISC_RING_BUFFER_ADDR);
+#else
+    return reinterpret_cast<uint32_t*>(RING_BUFFER_ADDR);
+#endif
+}
+
+void push_to_ring_buffer(uint32_t val) {
+    auto buf = get_debug_ring_buffer();
+    volatile tt_l1_ptr uint32_t* current_ptr = &reinterpret_cast<DebugRingBufMemLayout *>(buf)->current_ptr;
+    uint32_t* data = reinterpret_cast<DebugRingBufMemLayout *>(buf)->data;
+
+    uint32_t curr_idx = *current_ptr;
+    if (curr_idx == DEBUG_RING_BUFFER_STARTING_MAGIC)
+        curr_idx = 0;
+    *current_ptr  = (curr_idx + 1) % RING_BUFFER_ELEMENTS;
```

Comment:
test for wrap by >= rather than mod (no HW support for mod, so turns into a fn call)